design-proposals/klip-28-create-or-replace.md

```diff
@@ -0,0 +1,200 @@
+existing data. For stream processing applications, a user may want to modify their application as 
+a result of:
+
+- Business Requirements: requirements simply change over time
+- Schema Evolution: the incoming data or required output has been modified
+- Optimizations: the same application can be executed more efficiently (either by user or engine)
+
+At time of writing, ksqlDB provides a crude mechanism for altering its application behavior: dropping 
+a query and restarting it at the earliest or latest offset. While this often works well in development 
+environments, there are limitations to its applicability in production:
+
+- Data Retention: the earliest available offset may not correspond with the beginning of time
+- Downtime: the delta between terminating and catch-up may be out of the application’s SLO
+- Output Routing: populating to the old output topic will produce duplicates while using a new output topic will require cascading migrations
+- Compute: recomputing the complete history for a query may not be feasible
+
+Kafka Streams provides more granular mechanisms (e.g. restarting queries with different behaviors 
+but identical consumer groups), but these methods burden users with extra complexity and lack guardrails.
+
+## Scope
+
+To better understand the scope of this KLIP and any future improvements, we define a taxonomy on 
+query upgrades as any combination of three types of characteristics: **source query, upgrade** and
+(optionally) **environment**:
+
+| **Category** | **Characteristic** | **Description** |
+|----------|----------------|-------------|
+| Query | Stateful | Stateful queries maintain local storage |
+| | Windowed | Windowed queries maintain a limited amount of state specified by a window in time
+| | Joined | Joined queries read from multiple sources
+| | Multistage | Multistage queries contain intermediate, non-user visible topics in Kafka
+| | Nondeterministic | Nondeterministic queries may produce different results when executing identical input
+| | Simple | Queries with none of the above characteristics
+| Upgrade | Transparent | Transparent upgrades change the way something is computer (e.g. improving a UDF performance)
+| | Data Selection | Data selecting query upgrades change which/how many events are emitted
+| | Schema Evolution | Schema evolving query upgrades change the output type of the data |
+| | Source Modifying | These upgrades change the source data, whether by means of modifying a JOIN or swapping out a source |
+| | Topology | These upgrades are invisible to the user, but change the topology, such as the number of sub-topologies or the ordering of operations (e.g. filter push down) |
+| | Scaling | Scaling upgrades change the physical properties of the query in order to enable better performance characteristics. |
+| | Unsupported | Unsupported upgrades are ones that will semantically change the query in an unsupported way. There are no plans to implement these migrations. |
```

Comment:
"No plans to implement these migrations" at this time, or ever?

Reply:
Ever - these are things like changing a stream to table, or a table to a stream. Things that just don't make sense in our mental model.